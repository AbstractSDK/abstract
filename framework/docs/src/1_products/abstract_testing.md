# Abstract-Testing

!todo

https://github.com/AbstractSDK/contracts/tree/main/packages/abstract-testing