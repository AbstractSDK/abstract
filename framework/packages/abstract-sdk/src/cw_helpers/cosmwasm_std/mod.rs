mod abstract_attributes;
mod wasm_query;

pub use abstract_attributes::AbstractAttributes;
pub use wasm_query::wasm_raw_query;
pub(crate) use wasm_query::ApiQuery;
