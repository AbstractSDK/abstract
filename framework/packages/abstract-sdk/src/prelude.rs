//! # SDK Prelude
//!
//! Re-exports all the API traits to make it easier to access them.
//!
//! ## Usage
//!
//! ```rust
//! use abstract_sdk::prelude::*;
//! ```

pub use crate::apis::{
    accounting::*, adapter::*, app::*, bank::*, execution::*, ibc::*, modules::*, respond::*,
    verify::*, version_registry::*,
};

#[cfg(feature = "stargate")]
pub use crate::apis::{distribution::*, stargate::feegrant::*};

pub use crate::ans_resolve::Resolve;
