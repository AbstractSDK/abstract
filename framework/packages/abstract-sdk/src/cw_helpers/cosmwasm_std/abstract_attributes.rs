use abstract_std::ABSTRACT_EVENT_TYPE;
use cosmwasm_std::{Attribute, Event, Response};

/// Trait for adding `abstract` attributes to something
pub trait AbstractAttributes {
    /// Add the provided `attrs`
    fn add_abstract_attributes<A: Into<Attribute>>(
        self,
        attrs: impl IntoIterator<Item = A>,
    ) -> Self;
}

/// Trait for adding `abstract` events to a [`Response`]
impl AbstractAttributes for Response {
    fn add_abstract_attributes<A: Into<Attribute>>(
        mut self,
        attrs: impl IntoIterator<Item = A>,
    ) -> Self {
        // Find the index of the first abstract event in the events vector
        let index = self.events.iter().position(|e| e.ty == ABSTRACT_EVENT_TYPE);

        if let Some(index) = index {
            // If an abstract event exists, replace it with a new event that has the additional attributes
            let event = self.events.remove(index);
            let new_event = event.add_attributes(attrs);
            self.events.insert(index, new_event);
        } else {
            // If an abstract event does not exist, create a new one with the additional attributes
            let new_event = Event::new(ABSTRACT_EVENT_TYPE).add_attributes(attrs);
            self.events.push(new_event);
        }

        self
    }
}

#[cfg(test)]
mod test {
    use super::*;

    use speculoos::prelude::*;

    #[test]
    fn test_add_abstract_attributes_no_abstract_event() {
        let mut response = Response::default();

        assert_that!(response.events).has_length(0);

        response = response.add_abstract_attributes(vec![("key1", "value1")]);

        assert_that!(response.events).has_length(1);
        let event = response.events.first().unwrap();
        assert_that!(event.ty).is_equal_to(ABSTRACT_EVENT_TYPE.to_string());
        assert_that!(event.attributes).has_length(1);

        let attr = event.attributes.first().unwrap();
        assert_that!(attr.key).is_equal_to("key1".to_string());
        assert_that!(attr.value).is_equal_to("value1".to_string());
    }

    #[test]
    fn test_add_abstract_attributes_with_abstract_event() {
        const KEY_1: &str = "abstract_key1";
        const VALUE_1: &str = "abstract_value1";
        let mut response = Response::default()
            .add_event(Event::new(ABSTRACT_EVENT_TYPE).add_attributes(vec![(KEY_1, VALUE_1)]));

        assert_that!(response.events).has_length(1);
        let event = response.events.first().unwrap();
        assert_that!(event.ty).is_equal_to(ABSTRACT_EVENT_TYPE.to_string());
        assert_that!(event.attributes).has_length(1);

        // Add the attributes
        const KEY_2: &str = "abstract_key2";
        const VALUE_2: &str = "abstract_value2";
        response = response.add_abstract_attributes(vec![(KEY_2, VALUE_2)]);

        assert_that!(response.events).has_length(1);
        let event = response.events.first().unwrap();
        assert_that!(event.ty).is_equal_to(ABSTRACT_EVENT_TYPE.to_string());
        // should now be 2
        assert_that!(event.attributes).has_length(2);

        let attr1 = event.attributes.first().unwrap();
        assert_that!(attr1.key).is_equal_to(KEY_1.to_string());
        assert_that!(attr1.value).is_equal_to(VALUE_1.to_string());

        let attr2 = event.attributes.get(1).unwrap();
        assert_that!(attr2.key).is_equal_to(KEY_2.to_string());
        assert_that!(attr2.value).is_equal_to(VALUE_2.to_string());
    }
}
