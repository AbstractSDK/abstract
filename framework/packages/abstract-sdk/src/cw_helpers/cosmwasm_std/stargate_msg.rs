use cosmos_sdk_proto::traits::Message;
use cosmwasm_std::to_binary;
use cosmwasm_std::CosmosMsg;
use cosmwasm_std::StdResult;
use serde::Serialize;

/// Shortcut helper as the construction of CosmosMsg::Stargate {...} can be quite verbose in contract code.
/// ```rust,norun
///         let msg = MsgCreateDenom {
///             sender: self.sender()?,
///             subdenom: self.subdenom.to_string(),
///         };
///
///         let msg = prost_stargate_msg(MsgCreateDenom::TYPE_URL, &msg)?;
/// ```
pub fn prost_stargate_msg(
    type_url: impl Into<String>,
    msg: impl Message,
) -> StdResult<cosmwasm_std::CosmosMsg> {
    Ok(cosmwasm_std::CosmosMsg::Stargate {
        type_url: type_url.into(),
        value: cosmwasm_std::Binary(msg.encode_to_vec()),
    })
}

#[cfg(test)]
mod test {
    use cosmos_sdk_proto::cosmos;

    use super::*;

    #[test]
    fn test_stargate_msg() {
        let msg = cosmos::feegrant::v1beta1::MsgRevokeAllowance {
            granter: "foo".to_owned(),
            grantee: "bar".to_owned(),
        };
        let _msg = prost_stargate_msg("/cosmos.feegrant.v1beta1.MsgGrantAllowance", &msg).unwrap();
    }
}
