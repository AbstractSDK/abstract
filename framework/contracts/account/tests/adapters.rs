use abstract_adapter::{
    mock::{self, MockError, MockExecMsg, MockInitMsg},
    AdapterError,
};
use abstract_integration_tests::{
    add_mock_adapter_install_fee, create_default_account, init_mock_adapter, install_adapter,
    install_adapter_with_funds, mock_modules,
    mock_modules::adapter_1::{MockAdapterI1V1, MockAdapterI1V2},
    AResult,
};
use abstract_interface::*;
use abstract_std::{
    account::{AccountModuleInfo, ModuleInstallConfig},
    adapter::{AdapterBaseMsg, AdapterRequestMsg, BaseExecuteMsg, BaseQueryMsgFns},
    objects::{
        fee::FixedFee,
        module::{ModuleInfo, ModuleVersion, Monetization},
    },
    *,
};
use abstract_testing::{mock_bech32::mock_bech32_sender, prelude::*};
use cosmwasm_std::{coin, coins};
use cw_orch::prelude::*;
use mock_modules::{adapter_1, V1, V2};
use speculoos::{assert_that, result::ResultAssertions, string::StrAssertions};

#[test]
fn installing_one_adapter_should_succeed() -> AResult {
    let chain = MockBech32::new("mock");
    let sender = chain.sender_addr();
    let deployment = Abstract::deploy_on(chain.clone(), mock_bech32_sender(&chain))?;
    let account = create_default_account(&sender, &deployment)?;
    let staking_adapter = init_mock_adapter(chain.clone(), &deployment, None, account.id()?)?;
    install_adapter(&account, TEST_MODULE_ID)?;

    let modules = account.expect_modules(vec![staking_adapter.address()?.to_string()])?;

    assert_that(&modules[0]).is_equal_to(&AccountModuleInfo {
        address: staking_adapter.address()?,
        id: TEST_MODULE_ID.to_string(),
        version: cw2::ContractVersion {
            contract: TEST_MODULE_ID.into(),
            version: TEST_VERSION.into(),
        },
    });

    // Configuration is correct
    let adapter_config = staking_adapter.base_config()?;
    assert_that!(adapter_config).is_equal_to(adapter::AdapterConfigResponse {
        ans_host_address: deployment.ans_host.address()?,
        dependencies: vec![],
        version_control_address: deployment.version_control.address()?,
    });

    // no authorized addresses registered
    let authorized = staking_adapter.authorized_addresses(account.addr_str()?)?;
    assert_that!(authorized)
        .is_equal_to(adapter::AuthorizedAddressesResponse { addresses: vec![] });

    take_storage_snapshot!(chain, "install_one_adapter");

    Ok(())
}

#[test]
fn installing_one_adapter_without_fee_should_fail() -> AResult {
    let chain = MockBech32::new("mock");
    let sender = chain.sender_addr();
    chain.set_balance(&sender, coins(12, "ujunox"))?;
    let deployment = Abstract::deploy_on(chain.clone(), mock_bech32_sender(&chain))?;
    let account = create_default_account(&sender, &deployment)?;
    init_mock_adapter(chain.clone(), &deployment, None, account.id()?)?;
    add_mock_adapter_install_fee(
        &deployment,
        Monetization::InstallFee(FixedFee::new(&coin(45, "ujunox"))),
        None,
    )?;
    // TODO, match the exact error
    assert_that!(install_adapter(&account, TEST_MODULE_ID)).is_err();

    // TODO, match the exact error
    assert_that!(install_adapter_with_funds(
        &account,
        TEST_MODULE_ID,
        &coins(12, "ujunox")
    ))
    .is_err();

    Ok(())
}

#[test]
fn installing_one_adapter_with_fee_should_succeed() -> AResult {
    let chain = MockBech32::new("mock");
    Abstract::deploy_on(chain.clone(), mock_bech32_sender(&chain))?;
    abstract_integration_tests::account::installing_one_adapter_with_fee_should_succeed(
        chain.clone(),
    )?;
    take_storage_snapshot!(chain, "install_one_adapter_with_fee");
    Ok(())
}

#[test]
fn install_non_existent_adapterid_should_fail() -> AResult {
    let chain = MockBech32::new("mock");
    let sender = chain.sender_addr();
    let deployment = Abstract::deploy_on(chain.clone(), mock_bech32_sender(&chain))?;
    let account = create_default_account(&sender, &deployment)?;

    let res = install_adapter(&account, "lol:no_chance");

    assert_that!(res).is_err();
    Ok(())
}

#[test]
fn install_non_existent_version_should_fail() -> AResult {
    let chain = MockBech32::new("mock");
    let sender = chain.sender_addr();
    let deployment = Abstract::deploy_on(chain.clone(), mock_bech32_sender(&chain))?;
    let account = create_default_account(&sender, &deployment)?;
    init_mock_adapter(chain, &deployment, None, account.id()?)?;

    let res = account.install_module_version(
        TEST_MODULE_ID,
        ModuleVersion::Version("1.2.3".to_string()),
        Some(&Empty {}),
        &[],
    );

    // testtodo: check error
    assert_that!(res).is_err();

    Ok(())
}

#[test]
fn installation_of_duplicate_adapter_should_fail() -> AResult {
    let chain = MockBech32::new("mock");
    let sender = chain.sender_addr();
    let deployment = Abstract::deploy_on(chain.clone(), mock_bech32_sender(&chain))?;
    let account = create_default_account(&sender, &deployment)?;
    let staking_adapter = init_mock_adapter(chain, &deployment, None, account.id()?)?;

    install_adapter(&account, TEST_MODULE_ID)?;

    let modules = account.expect_modules(vec![staking_adapter.address()?.to_string()])?;

    // assert proxy module
    // check staking adapter
    assert_that(&modules[0]).is_equal_to(&AccountModuleInfo {
        address: staking_adapter.address()?,
        id: TEST_MODULE_ID.to_string(),
        version: cw2::ContractVersion {
            contract: TEST_MODULE_ID.into(),
            version: TEST_VERSION.into(),
        },
    });

    // install again
    let second_install_res = install_adapter(&account, TEST_MODULE_ID);
    assert_that!(second_install_res)
        .is_err()
        .matches(|e| e.to_string().contains("test-module-id"));

    account.expect_modules(vec![staking_adapter.address()?.to_string()])?;

    Ok(())
}

#[test]
fn reinstalling_adapter_should_be_allowed() -> AResult {
    let chain = MockBech32::new("mock");
    let sender = chain.sender_addr();
    let deployment = Abstract::deploy_on(chain.clone(), mock_bech32_sender(&chain))?;
    let account = create_default_account(&sender, &deployment)?;
    let staking_adapter = init_mock_adapter(chain.clone(), &deployment, None, account.id()?)?;

    install_adapter(&account, TEST_MODULE_ID)?;

    let modules = account.expect_modules(vec![staking_adapter.address()?.to_string()])?;

    // check staking adapter
    assert_that(&modules[0]).is_equal_to(&AccountModuleInfo {
        address: staking_adapter.address()?,
        id: TEST_MODULE_ID.to_string(),
        version: cw2::ContractVersion {
            contract: TEST_MODULE_ID.into(),
            version: TEST_VERSION.into(),
        },
    });

    // uninstall
    account.uninstall_module(TEST_MODULE_ID.to_string())?;

    // None expected
    account.expect_modules(vec![])?;

    // reinstall
    install_adapter(&account, TEST_MODULE_ID)?;

    account.expect_modules(vec![staking_adapter.address()?.to_string()])?;
    take_storage_snapshot!(chain, "reinstalling_adapter_should_be_allowed");

    Ok(())
}

/// Reinstalling the Adapter should install the latest version
#[test]
fn reinstalling_new_version_should_install_latest() -> AResult {
    let chain = MockBech32::new("mock");
    let sender = chain.sender_addr();
    let deployment = Abstract::deploy_on(chain.clone(), mock_bech32_sender(&chain))?;
    let account = create_default_account(&sender, &deployment)?;
    deployment
        .version_control
        .claim_namespace(TEST_ACCOUNT_ID, "tester".to_string())?;

    let adapter1 = MockAdapterI1V1::new_test(chain.clone());
    adapter1
        .deploy(V1.parse().unwrap(), MockInitMsg {}, DeployStrategy::Try)
        .unwrap();

    install_adapter(&account, &adapter1.id())?;

    let modules = account.expect_modules(vec![adapter1.address()?.to_string()])?;

    // check staking adapter
    assert_that(&modules[0]).is_equal_to(&AccountModuleInfo {
        address: adapter1.address()?,
        id: adapter1.id(),
        version: cw2::ContractVersion {
            contract: adapter1.id(),
            version: V1.into(),
        },
    });

    // uninstall tendermint staking
    account.uninstall_module(adapter1.id())?;

    account.expect_modules(vec![])?;

    let old_adapter_addr = adapter1.address()?;

    let adapter2 = MockAdapterI1V2::new_test(chain.clone());

    adapter2
        .deploy(V2.parse().unwrap(), MockInitMsg {}, DeployStrategy::Try)
        .unwrap();

    // check that the latest staking version is the new one
    let latest_staking = deployment
        .version_control
        .module(ModuleInfo::from_id_latest(&adapter1.id())?)?;
    assert_that!(latest_staking.info.version).is_equal_to(ModuleVersion::Version(V2.to_string()));

    // reinstall
    install_adapter(&account, &adapter2.id())?;

    let modules = account.expect_modules(vec![adapter2.address()?.to_string()])?;

    assert_that!(modules[0]).is_equal_to(&AccountModuleInfo {
        // the address stored for MockAdapterI was updated when we instantiated the new version, so this is the new address
        address: adapter2.address()?,
        id: adapter2.id(),
        version: cw2::ContractVersion {
            contract: adapter2.id(),
            // IMPORTANT: The version of the contract did not change although the version of the module in version control did.
            // Beware of this distinction. The version of the contract is the version that's imbedded into the contract's wasm on compilation.
            version: V2.to_string(),
        },
    });
    // assert that the new staking adapter has a different address
    assert_ne!(old_adapter_addr, adapter2.address()?);

    assert_that!(modules[0].address).is_equal_to(adapter2.as_instance().address()?);
    take_storage_snapshot!(chain, "reinstalling_new_version_should_install_latest");

    Ok(())
}

// struct TestModule = AppContract

#[test]
fn unauthorized_exec() -> AResult {
    let chain = MockBech32::new("mock");
    let sender = chain.sender_addr();
    let unauthorized = chain.addr_make("unauthorized");
    let deployment = Abstract::deploy_on(chain.clone(), mock_bech32_sender(&chain))?;
    let account = create_default_account(&sender, &deployment)?;
    let staking_adapter = init_mock_adapter(chain.clone(), &deployment, None, account.id()?)?;
    install_adapter(&account, TEST_MODULE_ID)?;
    // non-authorized address cannot execute
    let res = staking_adapter
        .call_as(&unauthorized)
        .execute(&MockExecMsg {}.into(), &[])
        .unwrap_err();
    assert_that!(res.root().to_string()).contains(format!(
        "Sender: {} of request to tester:test-module-id is not a Manager or Authorized Address",
        unauthorized
    ));
    // neither can the ROOT directly
    let res = staking_adapter
        .execute(&MockExecMsg {}.into(), &[])
        .unwrap_err();
    assert_that!(&res.root().to_string()).contains(format!(
        "Sender: {} of request to tester:test-module-id is not a Manager or Authorized Address",
        chain.sender_addr()
    ));
    Ok(())
}

#[test]
fn manager_adapter_exec() -> AResult {
    let chain = MockBech32::new("mock");
    let sender = chain.sender_addr();
    let deployment = Abstract::deploy_on(chain.clone(), mock_bech32_sender(&chain))?;
    let account = create_default_account(&sender, &deployment)?;
    let _staking_adapter_one = init_mock_adapter(chain.clone(), &deployment, None, account.id()?)?;

    install_adapter(&account, TEST_MODULE_ID)?;

    chain.set_balance(&account.address()?, vec![Coin::new(100_000u128, TTOKEN)])?;

    account.execute_on_module(
        TEST_MODULE_ID,
        Into::<abstract_std::adapter::ExecuteMsg<MockExecMsg>>::into(MockExecMsg {}),
    )?;

    Ok(())
}

#[test]
fn installing_specific_version_should_install_expected() -> AResult {
    let chain = MockBech32::new("mock");
    let sender = chain.sender_addr();
    let deployment = Abstract::deploy_on(chain.clone(), mock_bech32_sender(&chain))?;
    let account = create_default_account(&sender, &deployment)?;
    deployment
        .version_control
        .claim_namespace(TEST_ACCOUNT_ID, "tester".to_string())?;

    let adapter1 = MockAdapterI1V1::new_test(chain.clone());
    adapter1
        .deploy(V1.parse().unwrap(), MockInitMsg {}, DeployStrategy::Try)
        .unwrap();

    let v1_adapter_addr = adapter1.address()?;

    let adapter2 = MockAdapterI1V2::new_test(chain.clone());

    adapter2
        .deploy(V2.parse().unwrap(), MockInitMsg {}, DeployStrategy::Try)
        .unwrap();

    let expected_version = "1.0.0".to_string();

    // install specific version
    account.install_module_version(
        &adapter1.id(),
        ModuleVersion::Version(expected_version),
        Some(&MockInitMsg {}),
        &[],
    )?;

    let modules = account.expect_modules(vec![v1_adapter_addr.to_string()])?;
    let installed_module: AccountModuleInfo = modules[0].clone();
    assert_that!(installed_module.id).is_equal_to(adapter1.id());
    take_storage_snapshot!(chain, "installing_specific_version_should_install_expected");

    Ok(())
}

#[test]
fn account_install_adapter() -> AResult {
    let chain = MockBech32::new("mock");
    let sender = chain.sender_addr();
    let deployment = Abstract::deploy_on(chain.clone(), mock_bech32_sender(&chain))?;
    let account = create_default_account(&sender, &deployment)?;

    deployment
        .version_control
        .claim_namespace(TEST_ACCOUNT_ID, "tester".to_owned())?;

    let adapter = MockAdapterI1V1::new_test(chain.clone());
    adapter.deploy(V1.parse().unwrap(), MockInitMsg {}, DeployStrategy::Try)?;
    let adapter_addr = account.install_adapter(&adapter, &[])?;
    let module_addr = account
        .module_info(adapter_1::MOCK_ADAPTER_ID)?
        .unwrap()
        .address;
    assert_that!(adapter_addr).is_equal_to(module_addr);
    take_storage_snapshot!(chain, "account_install_adapter");
    Ok(())
}

#[test]
fn account_adapter_ownership() -> AResult {
    let chain = MockBech32::new("mock");
    let admin = mock_bech32_sender(&chain);
    let sender = chain.sender();

    let deployment = Abstract::deploy_on(chain.clone(), admin.clone())?;
    let account = create_default_account(&sender, &deployment)?;

    deployment
        .version_control
        .claim_namespace(TEST_ACCOUNT_ID, "tester".to_owned())?;

    let adapter = MockAdapterI1V1::new_test(chain.clone());
    adapter.deploy(V1.parse().unwrap(), MockInitMsg {}, DeployStrategy::Try)?;
    account.install_adapter(&adapter, &[])?;

    let proxy_addr = account.address()?;

    // Checking module requests

    // Can call either by account owner or manager
    adapter.call_as(&sender).execute(
        &mock::ExecuteMsg::Module(AdapterRequestMsg {
            account_address: Some(proxy_addr.to_string()),
            request: MockExecMsg {},
        }),
        &[],
    )?;
    adapter.call_as(&account.address()?).execute(
        &mock::ExecuteMsg::Module(AdapterRequestMsg {
            account_address: Some(proxy_addr.to_string()),
            request: MockExecMsg {},
        }),
        &[],
    )?;

    // Not admin or manager
    let who = chain.addr_make("who");
    let err: MockError = adapter
        .call_as(&who)
        .execute(
            &mock::ExecuteMsg::Module(AdapterRequestMsg {
                account_address: Some(proxy_addr.to_string()),
                request: MockExecMsg {},
            }),
            &[],
        )
        .unwrap_err()
        .downcast()
        .unwrap();
    assert_eq!(
        err,
        MockError::Adapter(AdapterError::UnauthorizedAddressAdapterRequest {
            adapter: adapter_1::MOCK_ADAPTER_ID.to_owned(),
            sender: who.to_string()
        })
    );

    // Checking base requests

    // Can call either by account owner or manager
    adapter.call_as(&sender).execute(
        &mock::ExecuteMsg::Base(BaseExecuteMsg {
            account_address: Some(proxy_addr.to_string()),
            msg: AdapterBaseMsg::UpdateAuthorizedAddresses {
                to_add: vec![chain.addr_make("123").to_string()],
                to_remove: vec![],
            },
        }),
        &[],
    )?;
    adapter.call_as(&account.address()?).execute(
        &mock::ExecuteMsg::Base(BaseExecuteMsg {
            account_address: Some(proxy_addr.to_string()),
            msg: AdapterBaseMsg::UpdateAuthorizedAddresses {
                to_add: vec![chain.addr_make("234").to_string()],
                to_remove: vec![],
            },
        }),
        &[],
    )?;

    // Not admin or manager
    let err: MockError = adapter
        .call_as(&who)
        .execute(
            &mock::ExecuteMsg::Base(BaseExecuteMsg {
                account_address: Some(proxy_addr.to_string()),
                msg: AdapterBaseMsg::UpdateAuthorizedAddresses {
                    to_add: vec![chain.addr_make("345").to_string()],
                    to_remove: vec![],
                },
            }),
            &[],
        )
        .unwrap_err()
        .downcast()
        .unwrap();
    assert_eq!(
        err,
        MockError::Adapter(AdapterError::UnauthorizedAdapterRequest {
            adapter: adapter_1::MOCK_ADAPTER_ID.to_owned(),
            sender: who.to_string()
        })
    );

    Ok(())
}

#[test]
fn subaccount_adapter_ownership() -> AResult {
    let chain = MockBech32::new("mock");
    let sender = chain.sender_addr();
    let deployment = Abstract::deploy_on(chain.clone(), mock_bech32_sender(&chain))?;
    let account = create_default_account(&sender, &deployment)?;

    deployment
        .version_control
        .claim_namespace(TEST_ACCOUNT_ID, "tester".to_owned())?;

    let adapter = MockAdapterI1V1::new_test(chain.clone());
    adapter.deploy(V1.parse().unwrap(), MockInitMsg {}, DeployStrategy::Try)?;

    let sub_account = account.create_and_return_sub_account(
        AccountDetails {
            name: "My subaccount".to_string(),
            description: None,
            link: None,
            namespace: None,
            install_modules: vec![ModuleInstallConfig::new(
                ModuleInfo::from_id_latest(adapter_1::MOCK_ADAPTER_ID).unwrap(),
                None,
            )],
            account_id: None,
        },
        &[],
    )?;

    let module = sub_account
        .module_info(adapter_1::MOCK_ADAPTER_ID)?
        .unwrap();
    adapter.set_address(&module.address);

    let proxy_addr = sub_account.address()?;

    // Checking module requests

    // Can call either by account owner or manager
    adapter.call_as(&sender).execute(
        &mock::ExecuteMsg::Module(AdapterRequestMsg {
            account_address: Some(proxy_addr.to_string()),
            request: MockExecMsg {},
        }),
        &[],
    )?;
    adapter.call_as(&sub_account.address()?).execute(
        &mock::ExecuteMsg::Module(AdapterRequestMsg {
            account_address: Some(proxy_addr.to_string()),
            request: MockExecMsg {},
        }),
        &[],
    )?;

    // Not admin or manager
    let who = chain.addr_make("who");
    let err: MockError = adapter
        .call_as(&who)
        .execute(
            &mock::ExecuteMsg::Module(AdapterRequestMsg {
                account_address: Some(proxy_addr.to_string()),
                request: MockExecMsg {},
            }),
            &[],
        )
        .unwrap_err()
        .downcast()
        .unwrap();
    assert_eq!(
        err,
        MockError::Adapter(AdapterError::UnauthorizedAddressAdapterRequest {
            adapter: adapter_1::MOCK_ADAPTER_ID.to_owned(),
            sender: who.to_string()
        })
    );

    // Checking base requests

    // Can call either by account owner or manager
    adapter.call_as(&sender).execute(
        &mock::ExecuteMsg::Base(BaseExecuteMsg {
            account_address: Some(proxy_addr.to_string()),
            msg: AdapterBaseMsg::UpdateAuthorizedAddresses {
                to_add: vec![chain.addr_make("123").to_string()],
                to_remove: vec![],
            },
        }),
        &[],
    )?;
    adapter.call_as(&sub_account.address()?).execute(
        &mock::ExecuteMsg::Base(BaseExecuteMsg {
            account_address: Some(proxy_addr.to_string()),
            msg: AdapterBaseMsg::UpdateAuthorizedAddresses {
                to_add: vec![chain.addr_make("234").to_string()],
                to_remove: vec![],
            },
        }),
        &[],
    )?;

    // Not admin or manager
    let err: MockError = adapter
        .call_as(&who)
        .execute(
            &mock::ExecuteMsg::Base(BaseExecuteMsg {
                account_address: Some(proxy_addr.to_string()),
                msg: AdapterBaseMsg::UpdateAuthorizedAddresses {
                    to_add: vec![chain.addr_make("345").to_string()],
                    to_remove: vec![],
                },
            }),
            &[],
        )
        .unwrap_err()
        .downcast()
        .unwrap();
    assert_eq!(
        err,
        MockError::Adapter(AdapterError::UnauthorizedAdapterRequest {
            adapter: adapter_1::MOCK_ADAPTER_ID.to_owned(),
            sender: who.to_string()
        })
    );
    Ok(())
}
