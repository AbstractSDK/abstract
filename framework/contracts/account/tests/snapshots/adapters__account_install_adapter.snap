---
source: contracts/account/tests/adapters.rs
expression: all_storage
---
"abstract:account-local-0":
  - - aa
    - "false"
  - - ab
    - "{\"name\":\"Default Abstract Account\"}"
  - - af
    - "[]"
  - - ag
    - "{\"trace\":\"local\",\"seq\":0}"
  - - ai
    - "[]"
  - - contract_info
    - "{\"contract\":\"abstract:account\",\"version\":\"0.24.0\"}"
  - - ownership
    - "{\"owner\":{\"monarchy\":{\"monarch\":\"mock14cl2dthqamgucg9sfvv4relp3aa83e40yfzzc8\"}},\"pending_owner\":null,\"pending_expiry\":null}"
"abstract:account-local-1":
  - - "\u0000\u0002actester:mock-adapter1"
    - "\"mock14xc5dkz0rn8j99lxz69mkv3wzawmadg7xurkzy49m9yefmqx5c6s4lc5ql\""
  - - aa
    - "false"
  - - ab
    - "{\"name\":\"Default Abstract Account\"}"
  - - af
    - "[\"mock14xc5dkz0rn8j99lxz69mkv3wzawmadg7xurkzy49m9yefmqx5c6s4lc5ql\"]"
  - - ag
    - "{\"trace\":\"local\",\"seq\":1}"
  - - ah
    - "[[{\"info\":{\"namespace\":\"tester\",\"name\":\"mock-adapter1\",\"version\":{\"version\":\"1.0.0\"}},\"reference\":{\"adapter\":\"mock14xc5dkz0rn8j99lxz69mkv3wzawmadg7xurkzy49m9yefmqx5c6s4lc5ql\"}},null]]"
  - - ai
    - "[]"
  - - contract_info
    - "{\"contract\":\"abstract:account\",\"version\":\"0.24.0\"}"
  - - ownership
    - "{\"owner\":{\"monarchy\":{\"monarch\":\"mock1pgm8hyk0pvphmlvfjc8wsvk4daluz5tgrw6pu5mfpemk74uxnx9qwrtv4f\"}},\"pending_owner\":null,\"pending_expiry\":null}"
"abstract:ans-host":
  - - be
    - "[]"
  - - cfg
    - "{\"next_unique_pool_id\":1}"
  - - contract_info
    - "{\"contract\":\"abstract:ans-host\",\"version\":\"0.24.0\"}"
  - - ownership
    - "{\"owner\":\"mock14cl2dthqamgucg9sfvv4relp3aa83e40yfzzc8\",\"pending_owner\":null,\"pending_expiry\":null}"
"abstract:ibc-client":
  - - contract_info
    - "{\"contract\":\"abstract:ibc-client\",\"version\":\"0.24.0\"}"
  - - mod
    - "{\"module\":\"abstract:ibc-client\",\"version\":\"0.24.0\",\"dependencies\":[],\"metadata\":null}"
  - - ownership
    - "{\"owner\":\"mock14cl2dthqamgucg9sfvv4relp3aa83e40yfzzc8\",\"pending_owner\":null,\"pending_expiry\":null}"
"abstract:ibc-host":
  - - contract_info
    - "{\"contract\":\"abstract:ibc-host\",\"version\":\"0.24.0\"}"
  - - ownership
    - "{\"owner\":\"mock14cl2dthqamgucg9sfvv4relp3aa83e40yfzzc8\",\"pending_owner\":null,\"pending_expiry\":null}"
"abstract:module-factory":
  - - contract_info
    - "{\"contract\":\"abstract:module-factory\",\"version\":\"0.24.0\"}"
  - - ownership
    - "{\"owner\":\"mock14cl2dthqamgucg9sfvv4relp3aa83e40yfzzc8\",\"pending_owner\":null,\"pending_expiry\":null}"
"abstract:registry":
  - - "\u0000\u0002cb\u0000\u0006tester\u0000\rmock-adapter11.0.0"
    - "{\"adapter\":\"mock14xc5dkz0rn8j99lxz69mkv3wzawmadg7xurkzy49m9yefmqx5c6s4lc5ql\"}"
  - - "\u0000\u0002cb\u0000\babstract\u0000\u0007account0.24.0"
    - "{\"account\":5}"
  - - "\u0000\u0002cb\u0000\babstract\u0000\bans-host0.24.0"
    - "{\"native\":\"mock1wx0qjtlz799pfxl73y2f4dv28nukztpmq2ztavk60v534tat9cds0pnytz\"}"
  - - "\u0000\u0002cb\u0000\babstract\u0000\bibc-host0.24.0"
    - "{\"native\":\"mock1ldaf3e9zxja5mrcw3clcu873wggmqhwfvqw9djjcjxxjwv59crqsc6n7r3\"}"
  - - "\u0000\u0002cb\u0000\babstract\u0000\bregistry0.24.0"
    - "{\"native\":\"mock1cag6cwygef6fuddgq2l44py7crpscufrza4mt3rum2axueemenmqapxshc\"}"
  - - "\u0000\u0002cb\u0000\babstract\u0000\nibc-client0.24.0"
    - "{\"native\":\"mock1g86ckugm7v9vvpadq9szg6rlf72pdl4a6n0d8apltq5t0grecpwqtrcaj0\"}"
  - - "\u0000\u0002cb\u0000\babstract\u0000\u000emodule-factory0.24.0"
    - "{\"native\":\"mock1vnj802lyyupfaez73w7axfs3xveraxen370vhcnumaufj29r5rxqwxdk4n\"}"
  - - "\u0000\u0002ch\u0000\u0005local\u0000\u0000\u0000\u0000"
    - "\"mock1ys8clus4qf2haemyck0uljz74dj2k8hu4ggfpe7vty6zd5j2982q9sf90k\""
  - - "\u0000\u0002ch\u0000\u0005local\u0000\u0000\u0000\u0001"
    - "\"mock1jdl4tasw0x00k2edyf8656lqxy95gx06h2nt5n8qjzv45yxr5ekq0myk0q\""
  - - "\u0000\u0005bbbmbabstract"
    - "{\"trace\":\"local\",\"seq\":0}"
  - - "\u0000\u0005bbbmbtester"
    - "{\"trace\":\"local\",\"seq\":1}"
  - - "\u0000\bblblblbl\u0000\u0005local\u0000\u0000\u0000\u0001"
    - "\"tester\""
  - - cfg
    - "{\"security_disabled\":true,\"namespace_registration_fee\":null}"
  - - ci
    - "2"
  - - contract_info
    - "{\"contract\":\"abstract:registry\",\"version\":\"0.24.0\"}"
  - - ownership
    - "{\"owner\":\"mock14cl2dthqamgucg9sfvv4relp3aa83e40yfzzc8\",\"pending_owner\":null,\"pending_expiry\":null}"
"tester:mock-adapter1":
  - - base_state
    - "{}"
  - - contract_info
    - "{\"contract\":\"tester:mock-adapter1\",\"version\":\"1.0.0\"}"
  - - mod
    - "{\"module\":\"tester:mock-adapter1\",\"version\":\"1.0.0\",\"dependencies\":[],\"metadata\":null}"
